iModelCore/ECPresentation/Source/UpdateHandler.cpp

```diff
@@ -832,45 +713,101 @@ struct HierarchyRefresher
         {
         if (identifier.GetParentNodeKey().IsNull())
             return nullptr;
-        CombinedHierarchyLevelIdentifier const& levelIdentifier = identifier.GetHierarchyLevelIdentifier();
-        return NodesLocater(m_connection, m_nodesCache, m_providerFactory, m_providerContextFactory).LocateNode(*identifier.GetParentNodeKey(), levelIdentifier.GetRulesetId());
+
+        auto cachedParentNode = m_nodesCache->LocateNode(m_connection, identifier.GetHierarchyLevelIdentifier().GetRulesetId(), *identifier.GetParentNodeKey());
+        if (cachedParentNode.IsNull())
+            return nullptr;
+
+        return cachedParentNode;
         }
 
     /*---------------------------------------------------------------------------------**//**
     * @bsimethod
     +---------------+---------------+---------------+---------------+---------------+------*/
-    NavNodesProviderPtr RefreshNavNodesProvider(AffectedHierarchyLevelIdentifier const& identifier, NavNodeCP parentNode)
+    bool IsNodeVisible(NavNodeCR node, Utf8StringCR rulesetId, UiState const& hierarchyState) const
+        {
+        NavNodeCPtr virtualParentNode = m_nodesCache->GetNode(m_nodesCache->GetVirtualParentNodeId(node.GetNodeId()));
+        auto const& parentLevelState = hierarchyState.GetHierarchyLevelState(virtualParentNode.get());

```

Comment:
You're right about being able to track expansion for physical parents only. For virtual nodes we won't have any state, so they'll always appear as not expanded here, causing us to make an early return with `false` result. But that's actually what we want here - this function is called to handle the case when physical parent becomes virtual during parent hierarchy level update. If it does, then we don't want to report anything for this level, because it's already reported when we updated the parent level.